src/state/CallViewModel.ts

```diff
@@ -887,17 +891,59 @@ export class CallViewModel extends ViewModel {
     : NEVER;
 
   /**
-   * Emits whenever the RTC session tells us that it intends to ring for a given
-   * duration.
+   * Whenever the RTC session tells us that it intends to ring the remote
+   * participant's devices, this emits an Observable tracking the current state of
+   * that ringing process.
    */
-  private readonly beginRingingForMs$ = (
+  private readonly ring$: Observable<
+    Observable<"ringing" | "timeout" | "decline">
+  > = (
     fromEvent(
       this.matrixRTCSession,
       MatrixRTCSessionEvent.DidSendCallNotification,
-    ) as Observable<[IRTCNotificationContent]>
-  )
-    // event.lifetime is expected to be in ms
-    .pipe(map(([notificationEvent]) => notificationEvent?.lifetime ?? 0));
+    ) as Observable<
+      Parameters<
+        MatrixRTCSessionEventHandlerMap[MatrixRTCSessionEvent.DidSendCallNotification]
+      >
+    >
+  ).pipe(
+    filter(
+      ([notificationEvent]) => notificationEvent.notification_type === "ring",
+    ),
+    map(([notificationEvent]) => {
+      const lifetimeMs = notificationEvent?.lifetime ?? 0;
+      return concat(
+        lifetimeMs === 0
+          ? // If no lifetime, skip the ring state
+            EMPTY
+          : // Ring until lifetime ms have passed
+            timer(lifetimeMs).pipe(
+              ignoreElements(),
+              startWith("ringing" as const),
+            ),
+        // The notification lifetime has timed out, meaning ringing has likely
+        // stopped on all receiving clients.
+        of("timeout" as const),
+        NEVER,
+      ).pipe(
+        takeUntil(
+          (
+            fromEvent(this.matrixRoom, RoomEvent.Timeline) as Observable<
+              Parameters<EventTimelineSetHandlerMap[RoomEvent.Timeline]>
+            >
+          ).pipe(
+            filter(
+              ([event]) =>
```

Comment:
What is exactly the expected logic here?
- Are we trying to check if the call was declined by another of our device? and then stop ringing here
- Or is it that we are the caller and trying to see if one/all of the callee have declined?

So should we check the decline event sender? if we are in a DM or not?